File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        mono_stack = [[temperatures[0], 0]]
        n = len(temperatures)
        ans = [0] * n 
        for i in range(1, n):
            while mono_stack and temperatures[i] > mono_stack[-1][0]:            
                _, k = mono_stack.pop()
                ans[k] = i - k
            mono_stack.append([temperatures[i], i])
        return ans
        #n = len(temperatures)
        #ans = [0] * n
        #mono_stack = [[0, temperatures[0]]]
        #for i in range(1, n):
        #    while mono_stack and temperatures[i] > mono_stack[-1][1]:
        #        k, _ = mono_stack.pop()
        #        ans[k] = i - k
        #    mono_stack.append([i, temperatures[i]])
        #return ans

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size());
        stack<int> s;
        for (int i = 0; i < temperatures.size(); i++) {
            while (!s.empty() && temperatures[s.top()] < temperatures[i]) {
                res[s.top()] = i - s.top();
                s.pop();
            } 
            s.push(i);
        }
        return res;
    }
};
//class Solution {
//public:
//    vector<int> dailyTemperatures(vector<int>& temperatures) {
//        vector<int> ans;
//        int n = temperatures.size();
//        for (int i = 0; i < n; i++) {
//            for (int j = i+1; j < n; j++) {
//                if (temperatures[i] < temperatures[j]) {
//                    ans.push_back(j-i);
//                    break;
//                } 
//            }
//            if (ans.size() != i+1) ans.push_back(0);
//        }
//        return ans;
//    }
//};